### DISCUSSION TITLE
Enhance device meta data beyond name and location

### DISCUSSION THREAD

**Discussion by ahochsteger** [zwave-js/zwave-js-ui#598] [Ideas]
Currently it's possible to set a **location** and device **name** that can be used to derive data for different integrations (e.g. MQTT topics, Home Assistant discovery).
I think there's potential to do much more, esp. if you want to use it to drive automation.

Here's some food for thought to get the discussion started:
* **Unique name or ID:** Smart Home systems refer to a device using a unique name or ID that can be used for further automations (e.g. device name in Home Assistant).
* **Descriptive title:** This title can be used in UIs of z2m or Smart Home systems
* **Unique location name or ID:** Smart Home systems may identify locations (or areas) using a unique name or ID
* **Channel name:** Using `switch_light` or `switch_fan` instead of `switch_1` and `switch_2` would greatly improve the legibility of automations that refer to a certain channel. It also allows to decouple the exact channel number from automation rules - esp. if they may change.
* **Description:** A longer description that may be used to describe the purpose of the device or document certain device aspects.
* **Custom metadata:** Ability to define custom metadata that may be used e.g. in HASS discovery templates or MQTT topics

The current fields **name** and **location** already cover some of them but there's no clear semantic distinction and so one has to decide between the following ways:
* put the unique name there and suffer from non-human friendly device names in the UI of the smart home system
* put a human readable title there (and automatically derive the name from it), causing unwanted changes to internal device names if the UI title is changed

The same may apply to the location field.

Let's discuss what kind of metadata is needed and what use cases they are required for ...

**Comment by robertsLando** [comment]
I don't get the point of this, could you show me an example where this would be useful?

**Reply by ahochsteger** [reply]
@robertsLando  of course - these are some of my use cases:

**Device names:**
Currently I'm using `location` and `name` in a way that generates short, unique and stable device names in Home Assistant. Even if the node ID changes it would not really affect the Home Assistant side, because the device names are stable.
The problem is, that this way they are not very user friendly shown in the UI which hinders broader use by my family members:
![image](https://user-images.githubusercontent.com/207989/108165946-b9b5df00-70f3-11eb-8d28-5c13e617be30.png)
I assume that also other features are affected as well, e.g. integration with voice assists like Google Assistant.
Being able to manage metadata like a unique id **and** a human-readable device name would improve that as we could automatically use that in the generated discovery JSON.

**Grouping of devices:**
Currently I mostly create my dashboards and groups manually. Some cards are automatically populated with items using some HACS integrations but the possibility is limited to filtering by the device name, since this is the only metadata I can use from z2m.
Here I'd like to be able to group or filter them by additional metadata I could attach to devices, e.g.:
* Group all door/window sensors representing doors or windows to the outside
* Group all thermostats in sleeping rooms and control them together
This shows an example that uses (manually managed) device groups for thermostats to control multiple of them together:
![image](https://user-images.githubusercontent.com/207989/108168780-2501b000-70f8-11eb-8eff-323fb419c700.png)
I'm not yet sure, if additional (or custom) metadata can easily be provided to Home Assistent using automatic device discovery but at least one could use scripts to make use of that data to generate groups and dashboard configurations outside of Home Assistant.

**Named channels:**
I have a dual switch (TKB Home TZ74) which shows up like this on the Home Assistant tab:
![image](https://user-images.githubusercontent.com/207989/108165055-36e05480-70f2-11eb-8be4-7425ad22ba3a.png)
As you can see, the object IDs are called `switch_1` and `switch_2`.
To use them for automations on the Home Assistent side they show up like this:
![image](https://user-images.githubusercontent.com/207989/108165536-04832700-70f3-11eb-9074-4f57d3c0a383.png)
Which is not very user friendly, since you don't always remember which number you'll have to use esp. if you have a lot of dual switches in use.
Being able to set a unique name (and a human-readable title) to individual switch channels would improve that as we could automatically use that in the generated discovery JSON.

**Switch device types:**
Currently all switches show up as power switch - no matter, if they control lights, fans or something else.
Being able to attach the type to individual switch channels would improve that as we could automatically use that in the generated discovery JSON.

In general I'd like to put more data into `nodes.json` and use that data to create a more dynamically discovery configuration or use that data e.g. for more dynamic MQTT topic names.
Of course, if no of such device metadata is configured the current defaults would still be the same to stay backwards compatible.
For special edge cases, where it is not possible to transport this information to home assistant via discovery configuration one could still use that information to automate the configuration generation outside of Home Assistant.

**Reply by robertsLando** [reply]
> Currently all switches show up as power switch - no matter, if they control lights, fans or something else.
Being able to attach the type to individual switch channels would improve that as we could automatically use that in the generated discovery JSON.

This could be fixed by using the device class informations, btw atm @billiaz is supporting mqtt discovery, I'm not focused on that as now there is also hass with his zwavejs server implementation working on that side so it's not a priority.

About device names. AFAIK hass allows you to give entities a friendly name property, couldn't you just use that for all this stuff?

**Reply by ahochsteger** [reply]
> About device names. AFAIK hass allows you to give entities a friendly name property, couldn't you just use that for all this stuff?

I could use the `name` in z2m for the friendly name but this affects everything (from MQTT topic names to device names) and in turn breaks everything if the friendly name changes.
Another problem is that - depending on the language you use - there may be characters in the friendly name that are not suitable for unique device names and may affect other systems that use this data.

Maybe I'm missing something but I wonder if nobody else has faced these issues?

**Reply by robertsLando** [reply]
@billiaz, @DanielWinks ?

**Reply by ahochsteger** [reply]
Of course I could keep it the way I currently do it (use name as unique internal device name) and manually set the friendly device name in Home Assistent but that defeats the purpose of auto-discovery and since I'm using it not only for Home Assistent but also for OpenHAB as well I'd have to manually change the friendly names multiple times in multiple systems that use the devices provided by z2m.

**Reply by robertsLando** [reply]
what about improve the entity template for this? Maybe by allowing to use a function
